lib: comply to coding guidelines MISRA C:2012 Rule 14.4

simhein opened the pull request:
Partly fix coding guideline MISRA C:2012 Rule 14.4 in lib:
> The controlling expression of an if statement
> and the controlling expression of an iteration-statement shall have
> essentially Boolean type.

Use `while (true)' instead of `while (1)' to express infinite loops.
Use comparisons with zero instead of implicitly testing integers.
Use comparisons with NULL instead of implicitly testing pointers.
Use comparisons with NUL instead of implicitly testing plain chars.

This PR is part of the enhancement issue #48002 which port the coding guideline fixes done by BUGSENG on the https://github.com/zephyrproject-rtos/zephyr/tree/v2.7-auditable-branch back to main

The commit in this PR is a subset of the original auditable-branch commit:
5d02614e34a86b549c7707d3d9f0984bc3a5f22a

----

This PR also change the use of the bit-wise OR to the boolean OR operator for the 1-bit boolean bitfields in lib/os/cbprintf_complete.c

simhein committed:
lib: comply partly to coding guidelines MISRA C:2012 Rule 14.4

MISRA C:2012 Rule 14.4 (The controlling expression of an if statement
and the controlling expression of an iteration-statement shall have
essentially Boolean type.)

Use `while (true)' instead of `while (1)' to express infinite loops.
Use comparisons with zero instead of implicitly testing integers.
Use comparisons with NUL instead of implicitly testing plain chars.

The commit in this PR is a subset of the original auditable-branch commit:
5d02614e34a86b549c7707d3d9f0984bc3a5f22a

Signed-off-by: Simon Hein <SHein@baumer.com>

keith-packard commented on lib/os/cbprintf_complete.c:
```diff
@@ -1815,7 +1815,7 @@ int z_cbvprintf_impl(cbprintf_cb out, void *ctx, const char *fp,
 			OUTS(cp, bpe);
 		} else {
-			if (conv->altform_0c | conv->altform_0) {
+			if ((conv->altform_0c | conv->altform_0) != 0) {
```
would this be better as if (conv->altform_0c !=0 || conv->altform_0 != 0) ? (I haven't looked at the surrounding code)

simhein replied:
I think it would be more readable if we change it. But than i would recommend also doing if ((conv->altform_0c != false) || (conv->altform_0 != false )) because altform and altform_0c are of boolean type. I can live with both solutions what would be your recomendation?

keith-packard replied:
Sorry, I hadn't looked at the underlying types of the two values. As they're both boolean, then the correct code
would just be `if (conv->altform_0c || conv->altform_0)` -- switch from the bitwise to the boolean `or` operator.

simhein replied:
Thank you, I will adapt the commit

keith-packard reviewed (dismissed):
Looks good to me. One minor question, but nothing blocking a merge.

keith-packard commented:
The compliance check failure is caused by editing a macro which includes a 'return' statement, so it can probably be ignored.

simhein commented:
> The compliance check failure is caused by editing a macro which includes a 'return' statement, so it can probably be ignored.

Yes I looked also into this. Only the while ( 0 ) was changed to while ( true ) in this macro and nothing other was changed, it seems that this was ignored before or was implemented before the check was in the CI

keith-packard reviewed (dismissed):
This looks great now, thanks for the update!

simhein commented:
@andyross can you please take a look? I feel like that the failing CI can be ignored. As Keith mentioned the compliance check failure is caused by editing a macro which includes a 'return' statement. I assume the check was introduced after the macro was implemented or was ignored before.

andyross commented on lib/os/cbprintf_complete.c:
```diff
@@ -1815,7 +1815,7 @@ int z_cbvprintf_impl(cbprintf_cb out, void *ctx, const char *fp,
 			OUTS(cp, bpe);
 		} else {
-			if (conv->altform_0c | conv->altform_0) {
+			if (conv->altform_0c || conv->altform_0) {
```
Are we 100% sure this was a typo and not testing for a non-zero result from bitwise OR?  Deserves an explicit callout in the commit, or to be a commit on its own.  This is a bugfix not a coding style update.

keith-packard replied:
yes, both of these values are 1-bit boolean bitfields, and so should be treated as boolean instead of binary.
I agree that a note in the commit would be good to clarify this.

simhein replied:
If this is really a bugfix, I think moving this change into a extra commit with its own PR would be more appropriate. Than it is not "hidden" in a coding guidelines adaption PR. What do you think?

keith-packard replied:
Could go either way - it's not fixing a execution bug as the bit-wise and boolean or operators end up producing the same effect when operating on two boolean values. I can also easily imagine someone writing the bitwise version because it was 'faster'. Whether you consider it a source-level bug seems like a matter of opinion when using the fine C language.

simhein replied:
That is true. In this case I will move that to an extra commit in this PR to have the explicit callout as andy said. I will also mention it in the PR description.

simhein replied:
@andyross @keith-packard I moved the change to an own commit and updated the PR description

andyross reviewed (dismissed):
Some nitpick pedantry about that "|" expression.

The checkpatch warning is indeed a false positive on pre-existing code.  It probably wouldn't be a bad idea to update the CONFIG_HEAP_VALIDATE stuff (it's pretty mature at this point, we probably don't need to maintain the giant whitebox rigging), but it's not relevant to this PR for sure.

simhein committed:
lib: os: Fix use of bit-wise OR for 1-bit boolean bitfields

Change the use of the bit-wise OR to the boolean OR operator
for the 1-bit boolean bitfields.

Signed-off-by: Simon Hein <SHein@baumer.com>
